pkg/typesystem/resolver.go

```diff
@@ -0,0 +1,80 @@
+package typesystem
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/oklog/ulid/v2"
+	"github.com/openfga/openfga/pkg/storage"
+)
+
+// TypesystemResolverFunc is a function that implementations can implement to provide lookup and
+// resolution of a Typesystem.
+type TypesystemResolverFunc func(ctx context.Context, storeID, modelID string) (*TypeSystem, error)
+
+// MemoizedTypesystemResolverFunc memoizes the provided TypesystemResolverFunc. It returns
+// the value from the first invocation of the underlying resolver function.
+//
+// The memoized resolver function is safe for concurrent use.
+func MemoizedTypesystemResolverFunc(fn TypesystemResolverFunc) TypesystemResolverFunc {
```

Comment:
This will need some adjustment, because the first time this is called without an authorization model id it will fetch the latest and memoize it. Subsequent calls won't fetch the latest. If `modelID` is an empty string we should bypass memoization.